Improve php-src docs sphinx build, also on *nix

hakre opened the pull request:
Close to a good share of a hand full of improvements of the php-src docs build.

Driven by my initial request in PHP-DEV for wiki editing right after which @cmb69  pointed me to the fact that there is more value in moving things towards php-src.

Therefore I looked into it and found `make html` not working addressed by adding a Makefile.

Found a benefit in having the requirements in a file instead of scattered, henceforth added a requirements.txt file and bound it in all places I could find. (the build could perhaps benefit from opening to branches other than master for docs/**, only publishing should remain on master branch only.)

Would be delighted to get a review by @iluuu1994 a) for the overall thing and b) there is a deprecation in the theme configuration and the theme is new to me.

cmb69 commented:
Thank you! This looks generally good to me (although I have *very* limited knowledge about Sphinx), but I'm slightly concerned that the build scripts for Windows and POSIX will diverge in the future. Unfortunately, there is not much common ground; maybe we should add some PHP to the mix? Another option would be to drop the Windows build script altogether; users could then use MSYS2 or Cygwin. That might be a hurdle for some, though.

hakre commented:
>  I'm slightly concerned that the build scripts for Windows and POSIX will diverge in the future. 

Yes, that is something to have under watch, always, but here I see little danger of that, as the build is driven by Python and Docutils (Sphinx) and is basically a simple command. I have spotted no additional magic in the configuration.

> maybe we should add some PHP to the mix?

No, not from my perspective. Preferable scripts should be written in Pyhton for a Docutils based build (Spinx).

>  That might be a hurdle for some, though.

With these requested changes the build instructions should work for users on both *nix and win systems, the overhead for that so far (one small batch and one small makefile) should not really bug us. And if it does I can offer to build back.

Build instructions remain streamlined:

    make html

and that's the most important for maintainability from my end.

iluuu1994 commented on docs/source/conf.py:
```diff
@@ -16,7 +16,6 @@
 author = 'The PHP Group'
 extensions = [
     'sphinx_design',
-    'sphinxawesome_theme.highlighting',
```
Why is this dropped?

hakre replied:
Deprecated, details are in the commit message @iluuu1994

iluuu1994 commented on docs/Makefile:
```diff
@@ -0,0 +1,29 @@
+html : $(SOURCEDIR)/.~
+	$(SPHINXBUILD) -M $@ $(SOURCEDIR) $(BUILDDIR)
+	@printf 'Browse the php-src docs \e]8;;%s\e\\%s\e]8;;\e\\ (or \e]8;;%s\e\\%s\e]8;;\e\\)\n' \
+		"file://$(abspath $(BUILDDIR))/$@/index.$@" "local" "$(GH_PAGES_URL)" "online"
```
Why do we spell out the online version? That makes it sound like it was deployed.

hakre replied:
Found it helpful to have the reference at hand. What would it make sound better to you? @iluuu1994

hakre replied:
Removed; deferred for a later improvement if any.

iluuu1994 commented on .github/workflows/docs.yml:
```diff
@@ -19,7 +19,7 @@ jobs:
-        run: pip install sphinx-design sphinxawesome-theme rstfmt
+        run: pip install -r docs/requirements.txt
       - name: Check formatting
         run: rstfmt --check -w 100 docs/source
```
This should probably use the make file, to make sure the commands stay consistent.

hakre replied:
The makefile does not install the python requirements.

iluuu1994 replied:
The commented line is the `rstfmt` command.

hakre replied:
That's an option, I'll give it a try. 

And a related question as I've learned over the days: The policy docs repository uses line length of 80 for .rst files, which is also the line-length of  .md here in php-src.. What about using -w 80 for rstfmt?

hakre replied:
This message got lost: That's certainly possible, and while we're at it: .rst is currently with line length 100 in php-src and .md with 80. In policy the rstfmt / .rst is with line length 80 as well. What about streamlining this and use a line length of 80 as well here in php-src for .rst? The PR is a good opportunity to do it. @iluuu1994

hakre replied:
Updated with the make binding in CI and also took in the 80 chars line length, please see the main thread of the PR @iluuu1994

iluuu1994 reviewed (commented):
Hi. Thanks for working on improving the docs.

Related: https://github.com/php/php-src/pull/16676

This can target master, we're not planning to maintain different versions of the document. Differences between versions can be referenced within the document itself.

> Another option would be to drop the Windows build script altogether

I agree with this. Note that I did not add this bat file, it ships by default in the sphinx template. If you're developing on Windows, wsl and cygwin seem like good options.

cmb69 commented:
> > Another option would be to drop the Windows build script altogether
> 
> I agree with this. Note that I did not add this bat file, it ships by default in the sphinx template. If you're developing on Windows, wsl and cygwin seem like good options.

Yeah, then please drop make.bat.

hakre commented:
> This can target master, we're not planning to maintain different versions of the document. Differences between versions can be referenced within the document itself.

Rebased.

> 
> > Another option would be to drop the Windows build script altogether
> 
> I agree with this. Note that I did not add this bat file, it ships by default in the sphinx template. If you're developing on Windows, wsl and cygwin seem like good options.

Dropped.

hakre commented:
Updated.

GH CI docs workflow now uses `make -C docs check-formatting` so the build is better managed.

Another late question of mine was to change .rst file line length to 80 characters, previously 100.

This is aligned with both .md files (the other documentation file format next to plain text) and with .rst files in [php/policies](/php/policies) (ref: php/policies@9a97c2fe64a2e50f019f05f88406221a96ebfac1) which uses the same tooling as `./docs` here.

Streamling here is beneficial I think, but unfortunately I'm not aware of the rationale behind a line length of 100. What I could gather from "upstream" (reStructuredText markup in the [Python Developer's Guide](https://devguide.python.org/)), 80 is also mentioned and the rstfmt utility shows the additional behavior when I build with it:

> The maximum line length is 80 characters for normal text, but tables, deeply indented code samples and long links may extend beyond that. ([ref](https://devguide.python.org/documentation/markup/#use-of-whitespace))

It is a trivial change and I previewed it locally. Building the HTML docs yielded the same objects (apart from some woff/2 ones that can be ignored in the test comparison).

Please let me know any objections.

hakre commented:
Thanks for fixing the CI MAC workflows @cmb69 !

iluuu1994 commented on docs/Makefile:
```diff
@@ -0,0 +1,36 @@
+	@printf 'Browse the \e]8;;%s\e\\%s\e]8;;\e\\.\n' \
+		"file://$(abspath $(BUILDDIR))/$@/index.$@" "php-src html docs locally"
+
+preflight : $(SOURCEDIR)/.~
```
I don't think building should depend on correct formatting. Some people might not like their code getting formatted on every build, depending on their IDE settings. I'd prefer just keeping formatting a manual command.

hakre replied:
They might not, however the build will do anyway (our assumption is that CI is parenting here). So the benefit then is to get the result in small changes as fast and often as possible. If some people do not like this, there is no requirement to use the managed build, they remain free to manage their build individually. There are no constraints enforced.

hakre replied:
And thinking: Looking into https://github.com/php/php-src/pull/16888#issuecomment-2491509834 , the build could benefit from a lint command for the same reasons as given there:

> maybe your CI should run sphinx-lint in addition to rstfmt. I'm quite sure sphinx-lint would have reported the broken literal markup at least.

iluuu1994 replied:
Fair enough. I still don't like forced execution for linters, but we can always remove if somebody complains.

iluuu1994 commented on .github/workflows/docs.yml:
```diff
@@ -19,9 +19,9 @@ jobs:
         uses: actions/checkout@v4
       - name: Install dependencies
-        run: pip install sphinx-design sphinxawesome-theme rstfmt
+        run: pip install -r docs/requirements.txt
```
We're assuming the user is in the docs directory.

hakre replied:
Are we? https://github.com/php/php-src/actions/runs/11872317688/job/33085877287#step:3:2 

Therefore, if so, then IMHO we should stop assuming and go with the reality that the CI build user is *not* (and never was) i the docs directory and CI run builds in-tree at work-tree root.

iluuu1994 reviewed (commented):
As mentioned, please revert to 100 chars per line. I don't see any good arguments for changing this. The rest looks reasonable. Thanks for working on it.

hakre commented:
> As mentioned, please revert to 100 chars per line. I don't see any good arguments for changing this. The rest looks reasonable. Thanks for working on it.

So what is the argument for 100 chars per line in the original rstfmt formatting suggestion? I could find no reference to it in all projects I'm personally aware of. Where does it originate from? What is its rationale? @iluuu1994

hakre commented:
rebased on target and removed the commit changing the line-length.

iluuu1994 commented:
Thank you @hakre!
